libc/dirent/readdir.c

```diff
@@ -0,0 +1,328 @@
+	rmdir(MAIN_DIR "/dir1/nest2");
+
+	rmdir(MAIN_DIR "/dir2/nest1");
+	rmdir(MAIN_DIR "/dir2/nest2");
+
+	rmdir(MAIN_DIR "/dir1");
+	rmdir(MAIN_DIR "/dir2");
+
+	remove(MAIN_DIR "/file1.txt");
+	remove(MAIN_DIR "/file2.dat");
+	remove(MAIN_DIR "/file3.json");
+
+	rmdir(MAIN_DIR);
+}
+
+
+TEST(dirent_readdir, long_name_directory_check)
+{
+	errno = 0;
+	DIR *dp = opendir(MAIN_DIR);
+	struct dirent *info;
+
+	char longDirName[NAME_MAX + 1] = { 0 };
+	longDirName[NAME_MAX] = '\0';
+
+	memset(longDirName, 'a', NAME_MAX);
+	char longDirPath[NAME_MAX + 2 + sizeof(MAIN_DIR)];
+	sprintf(longDirPath, MAIN_DIR "/%s", longDirName);
+
+	if (mkdir(longDirPath, 0777) == -1)
+		TEST_IGNORE_MESSAGE(strerror(errno));
+
+	int passed = 0;
+
+	while (info = readdir(dp)) {
+		if (!strcmp(longDirName, info->d_name)) {
+			closedir(dp);
+			rmdir(longDirPath);
+			TEST_PASS();
+		}
+	}
+
+	closedir(dp);
+	rmdir(longDirPath);
+	TEST_FAIL();
+}
+
+
+TEST(dirent_readdir, basic_listing_count)
+{
+	DIR *dp = opendir(MAIN_DIR);
+	int entry_counter = 0;
+	struct dirent *info;
+
+	if (!dp)
+		TEST_FAIL_MESSAGE(strerror(errno));
+
+	while (info = readdir(dp))
+		entry_counter++;
+
+	/* 6 files from setup, and both . and .. directories */
+	TEST_ASSERT_EQUAL(7, entry_counter);
+
+	closedir(dp);
+}
+
+
+TEST(dirent_readdir, reading_in_parent_and_child)
+{
+
+	DIR *dp1, *dp2;
+
+	dp1 = opendir(MAIN_DIR "/dir1");
+	dp2 = opendir(MAIN_DIR "/dir2");
+
+
+	TEST_ASSERT_NOT_NULL(readdir(dp1));
+	TEST_ASSERT_NOT_NULL(readdir(dp2));
+
+	/* After first readdir is done, removing contents from directory shall not influence output from readdir */
+
+	pid_t cid = fork();
+	/* Since there are two different dir streams, there is no reading from the same stream in two processes */
+
+	if (cid) {
+		/* Check for parent */
+		TEST_ASSERT_NOT_NULL(readdir(dp1));
+		closedir(dp1);
+		closedir(dp2);
+	}
+
+	else {
+		/* Check for child */
+		TEST_ASSERT_NOT_NULL(readdir(dp2));
+		closedir(dp1);
+		closedir(dp2);
+		kill(getpid(), SIGTERM);
+	}
+}
+
+
+TEST(dirent_readdir, symlink_inode_correct_number)
+{
+	mkdir("test", 0777);
+	mkdir("a1", 0777);
+	mkdir("a2", 0777);
+	symlink("test", "a3");
+
+	ino_t n1, n2;
+
+	struct dirent *info;
+	DIR *dp = opendir(".");
+
+	while (info = readdir(dp)) {
+		if (!strcmp("test", info->d_name)) {
+			n1 = info->d_ino;
+		}
+
+		if (!strcmp("a3", info->d_name)) {
+			n2 = info->d_ino;
+		}
+	}
+
+	closedir(dp);
+
+	TEST_ASSERT_NOT_EQUAL(n1, n2);
+
+	rmdir("test");
+	rmdir("a1");
+	rmdir("a2");
+	unlink("a3");
+}
+
+
+TEST(dirent_readdir, distinct_inode_nums)
+{
+	ino_t *inode_arr = calloc(INO_T_TEST_MAX_DIRS, sizeof(ino_t));
+	DIR *dp = opendir(MAIN_DIR);
+	struct dirent *info;
+	int inode_counter = 0;
+
+	/* assert distinct inodes */
+	while (info = readdir(dp)) {
+		TEST_ASSERT_EQUAL(0, d_ino_in(info->d_ino, inode_arr));
+		inode_arr[inode_counter++] = info->d_ino;
+	}
+	closedir(dp);
+	free(inode_arr);
+}
+
+
+TEST(dirent_readdir, same_file_reading_by_two_pointers)
```

Comment:
This test make no sense for me. Can you explain what it is about?

Reply:
This test case is about ensuring there is no interference between two distinct DIR pointers pointing to the same directory. They should be independent such that calling readdir() on one of them doesn't influence the state of the second one.
In this test case I call readdir() twice on one of the pointers, then do some operations on the second one, and then I make sure that the first one behaves as expected, in this case that it lists exact number of remaining directories.